TELCODOCS-1861: Lifecycle Agent Enhancements 4.17 (#80111)
openshift/openshift-docs: modules/ztp-image-based-upgrade-procedure-steps.adoc

// Module included in the following assemblies:
// * edge_computing/image-based-upgrade/ztp-image-based-upgrade.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-procedure-steps_{context}"]
= Performing an image-based upgrade on managed clusters at scale in several steps

> For use cases when you need better control of when the upgrade interrupts your service, you can upgrade a set of your managed clusters by using the `ImageBasedGroupUpgrade` CR with adding actions after the previous step is complete.

Reviewer: I'd avoid "previous" here and call out the section instead. In the event that somehow sections move, "previous" statements can cause issues.

Author: It's not referring to a procedure step but a step (one or more actions) in the upgrade plan, so the user will always have to evaluate the previous step in the upgrade plan.

<5> Upgrade plan details.
--

. Apply created file by running the following command on the hub cluster:
+
[source,terminal]
----
$ oc apply -f <filename>.yaml
----

. Monitor the status updates by running the following command on the hub cluster:
+
--
[source,terminal]
----
$ oc get ibgu -o yaml
----

.Example output
[source,yaml]
----
[...]
status:
  clusters:
  - completedActions:
    - action: Prep
    name: spoke1
  - completedActions:
    - action: Prep
    name: spoke4
  - failedActions:
    - action: Prep
    name: spoke6
[...]
----

The previous output of an example plan starts with the `Prep` stage only and you add actions to the plan based on the results of the previous step.
{cgu-operator} adds a label to the clusters to mark if the upgrade succeeded or failed.
For example, the `lcm.openshift.io/ibgu-prep-failed` is applied to clusters that failed the `Prep` stage.

After investigating the failure, you can add the `AbortOnFailure` step to your upgrade plan.
It moves the clusters labeled with `lcm.openshift.io/ibgu-<action>-failed` back to the `Idle` stage.
Any resources that are related to the upgrade on the selected clusters are deleted.
--

> . Optional: Add the `AbortOnFailure` action to your existing `ImageBasedGroupUpgrade` CR by running the following command:

Reviewer: Step 4 and 6 are the same. Is this intentional?

+
[source,terminal]
----
$ oc patch ibgu <filename> --type=json -p \
'[{"op": "add", "path": "/spec/plan/-", "value": {"actions": ["Upgrade"], "rolloutStrategy": {"maxConcurrency": 2, "timeout": 30}}}]'
----

. Optional: Add the `AbortOnFailure` action to your existing `ImageBasedGroupUpgrade` CR by running the following command:
+
[source,terminal]
----
$ oc patch ibgu <filename> --type=json -p \
'[{"op": "add", "path": "/spec/plan/-", "value": {"actions": ["AbortOnFailure"], "rolloutStrategy": {"maxConcurrency": 5, "timeout": 10}}}]'
----

.. Continue monitoring the status updates by running the following command:
+
[source,terminal]
----
$ oc get ibgu -o yaml
----

. Add the action to your existing `ImageBasedGroupUpgrade` CR by running the following command:
+
[source,terminal]
----
$ oc patch ibgu <filename> --type=json -p \
'[{"op": "add", "path": "/spec/plan/-", "value": {"actions": ["FinalizeUpgrade"], "rolloutStrategy": {"maxConcurrency": 10, "timeout": 3}}}]'
----

.Verification

* Monitor the status updates by running the following command:
+
--
[source,terminal]
----
$ oc get ibgu -o yaml
----

.Example output
[source,yaml]
----

> [...]

Reviewer: Replace `[...]` with `# ...`.

// Module included in the following assemblies:
// * edge_computing/image-based-upgrade/ztp-image-based-upgrade.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-image-based-upgrade-procedure-steps_{context}"]
= Performing an image-based upgrade on managed clusters at scale in several steps

Upgrade a set of your managed clusters by using the `ImageBasedGroupUpgrade` CR.
You can append an action after you evaluate the results of the previous steps.
This allows you to troubleshoot any failed stages throughout the upgrade procedure.

[IMPORTANT]
====
Ensure that the action combinations created by appending actions are supported.
For more information, see _Supported action combinations_.
====

.Prerequisites

* Log in to the hub cluster as a user with `cluster-admin` privileges.
* Create policies and `ConfigMap` objects for resources used in the image-based upgrade.
* Install {lcao} and OADP Operators on all managed clusters through the hub cluster.

.Procedure

> . Create a YAML file on the hub cluster that contains the `ImageBasedGroupUpgrade` CR, for example, named `example-group-upgrade.yaml`:

Reviewer 2: As already suggested in another file.

kind: ImageBasedGroupUpgrade
metadata:
  name: example-group-upgrade
  namespace: default
spec:
  clusterLabelSelectors: <1>
  - matchExpressions:
    - key: name
      operator: In
      values:
      - spoke1
      - spoke4
      - spoke6
  ibuSpec:
    seedImageRef: <2>
      image: quay.io/seed/image:4.16.0-rc.1
      version: 4.16.0-rc.1
      pullSecretRef:
        name: "<seed_pull_secret>"
  extraManifests: <3>
    - name: example-extra-manifests
      namespace: openshift-lifecycle-agent
  oadpContent: <4>
    - name: oadp-cm
      namespace: openshift-adp
  plan: <5>
  - actions: ["Prep"]
    rolloutStrategy:
      maxConcurrency: 2
      timeout: 2400
----
<1> Specify the set of clusters that you want to upgrade.
<2> Defines the target platform version, the seed image to be used, and the secret required to access the image.
<3> (Optional) Specify the list of `ConfigMap` resources that contain your custom catalog sources to retain after the upgrade, and your extra manifests to apply to the target cluster that are not part of the seed image.
<4> Specify the list of `ConfigMap` resources that contain the OADP `Backup` and `Restore` CRs.

> <5> Defines the upgrade plan.

Reviewer 2: Same feedback as for the callouts in the `modules/ztp-image-based-upgrade-procedure-one-step.adoc` file.

--

> . Apply the `example-group-upgrade.yaml` file by running the following command on the hub cluster:

Reviewer 2: Same feedback as in the `modules/ztp-image-based-upgrade-procedure-cancel.adoc` file:
```suggestion
. Apply the created file by running the following command on the hub cluster:
```

+
[source,terminal]
----

> $ oc apply -f example-group-upgrade.yaml

Reviewer 2: Same feedback as in the `modules/ztp-image-based-upgrade-procedure-cancel.adoc` file:
```suggestion
$ oc apply -f <filename>.yaml
```

----

. Monitor the status updates by running the following command on the hub cluster:
+
--
[source,terminal]
----
$ oc get ibgu -o yaml
----

.Example output
[source,yaml]
----
[...]
status:
  clusters:
  - completedActions:
    - action: Prep
    name: spoke1
  - completedActions:
    - action: Prep
    name: spoke4
  - failedActions:
    - action: Prep
    name: spoke6
[...]
----

> This example plan starts with the `Prep` stage only and you append actions to the plan based on the results of the step

Reviewer 2: `This example plan`--is this referring to the `.Example output` block that precedes this paragraph? If so, then consider adding this info as a callout. If not, then it's not clear what this is referring to.

Reviewer: "The previous example" would be clearer than the vague referent of "this". 

I think that the "After investigating..." paragraph can be moved to the next step? The block of text could be shortened and be more direct as the optional nature of the steps would gear the text around an action as against a detailed description.

{cgu-operator} adds a label to the clusters to mark if the upgrade succeeded or failed.
For example, the `lcm.openshift.io/ibgu-prep-failed` is applied to clusters that failed the `Prep` stage.

> After investigating the reason for failure, you can append the `AbortOnFailure` step to move the clusters labeled with `lcm.openshift.io/ibgu-<action>-failed` back to the `Idle` stage.

Reviewer 2: `append the `AbortOnFailure` step`--append  it where? In the CR? This is not clear to me, although I did notice some subsequent command block about this, and this step is also shown in the output block in the Verification, but I didn't get to the Verification yet, I'm still in the middle of this procedure.

Reviewer: +1 "step" is difficult to comprehend when the optional step states an "action".

Author: Again, I believe this should be clearer after reading the conceptual part - which was introduced in a separate PR:
https://80111--ocpdocs-pr.netlify.app/openshift-enterprise/latest/edge_computing/image_based_upgrade/ztp-image-based-upgrade#ztp-image-based-upgrade-supported-combinations_ztp-gitops

```
plan:
- actions: ["Prep", "Upgrade"]
  rolloutStrategy:
    maxConcurrency: 200
    timeout: 60
- actions: ["FinalizeUpgrade"]
  rolloutStrategy:
    maxConcurrency: 500
    timeout: 10```

`plan` is the upgrade plan
`- actions: {}` is one step that can have multiple actions
`["Prep"]` is one action 

You can have more than one action in one step to configure the `rolloutStrategy` for several actions, thus potentially reducing the upgrade time instead of doing it step-by-step. I'll clarify this in the one-shot vs step-by-step scenarios

> This deletes any resources on the selected clusters that are related to the upgrade and allows you to retry the upgrade after troubleshooting.

Reviewer 2: `This deletes any resources...`--this is not clear to me. What exactly (running a command or a setting a setting) deletes those resources?

Reviewer: "Allows you" is a form of anthrophomorphism.

> This deletes any resources on the selected clusters that are related to the upgrade so that you can retry the upgrade after troubleshooting.

Reviewer: +1 to Max about the "delete" action.

Author: Setting back to `Idle` is what triggers an Operator to delete the resources. [Context here](https://80111--ocpdocs-pr.netlify.app/openshift-enterprise/latest/edge_computing/image_based_upgrade/cnf-understanding-image-based-upgrade#cnf-image-based-upgrade-concept-idle_understanding-image-based-upgrade)

--

> . (Optional) Add the `AbortOnFailure` action to your existing `ImageBasedGroupUpgrade` CR by running the following command:

Reviewer 2: As mentioned elsewhere in this PR:
```suggestion
. Optional: Add the `AbortOnFailure` action to your existing `ImageBasedGroupUpgrade` CR by running the following command:
```

Reviewer: The [mod docs procedure template states](https://raw.githubusercontent.com/redhat-documentation/modular-docs/master/modular-docs-manual/files/TEMPLATE_PROCEDURE_doing-one-procedure.adoc): 
![Screenshot from 2024-09-20 16-29-31](https://github.com/user-attachments/assets/aa208628-9655-4fd9-8ef4-067fb88e7452)

+
[source,terminal]
----

> $ oc patch ibgu example-group-upgrade --type=json -p \

Reviewer 2: Half a suggestion (not sure what description should go inside the angle brackets):
```suggestion
$ oc patch ibgu <some_filename-agnostic_custom_resource_description> --type=json -p \
```

'[{"op": "add", "path": "/spec/plan/-", "value": {"actions": ["AbortOnFailure"], "rolloutStrategy": {"maxConcurrency": 5, "timeout": 10}}}]'
----

> .. Continue monitoring the status updates by running the following command:

Reviewer 2: Is this really a sub-step? If it is, then consider making the previous command in this step also a sub-step.

Reviewer: It might have been done this way to keep both actions in the same step block. If ".a" was converted to step 5, it would also be marked "Optional" but might be skipped by a customer assuming they could skip step 4 and go to 5. Maybe I'm wrong here.

+
[source,terminal]
----
$ oc get ibgu -o yaml
----

> . Add the action to your existing `ImageBasedGroupUpgrade` CR by running the following command:

Reviewer 2: `Add the action`--which action?

Reviewer: +1 Would be great to add a noun here to distinguish that "action" is some OCP component, such as a status, and not what a customer physically does.

Author: I would think this makes sense in the context of the feature: https://80111--ocpdocs-pr.netlify.app/openshift-enterprise/latest/edge_computing/image_based_upgrade/ztp-image-based-upgrade#ztp-image-based-upgrade-supported-combinations_ztp-gitops

We are using actions in the IBGU CR to step through the stages of the upgrade.
The reason I didn't specify which action we add here is because t's easier to reuse the lead sentence without specifying `Prep` or `Upgrade` or `FinalizeUpgrade` action. Let me know if you still want me to specify the actions

+
[source,terminal]
----
$ oc patch ibgu example-group-upgrade --type=json -p \
'[{"op": "add", "path": "/spec/plan/-", "value": {"actions": ["Upgrade"], "rolloutStrategy": {"maxConcurrency": 2, "timeout": 30}}}]'
----

> . (Optional) Add the `AbortOnFailure` action to your existing `ImageBasedGroupUpgrade` CR by running the following command:

Reviewer 2: As mentioned elsewhere in this PR:
```suggestion
. Optional: Add the `AbortOnFailure` action to your existing `ImageBasedGroupUpgrade` CR by running the following command:
```

+
[source,terminal]
----

> $ oc patch ibgu example-group-upgrade --type=json -p \

Reviewer 2: Half a suggestion (not sure what description should go inside the angle brackets):
```suggestion
$ oc patch ibgu <some_filename-agnostic_custom_resource_description> --type=json -p \
```

'[{"op": "add", "path": "/spec/plan/-", "value": {"actions": ["AbortOnFailure"], "rolloutStrategy": {"maxConcurrency": 5, "timeout": 10}}}]'
----

> .. Continue monitoring the status updates by running the following command:

Reviewer 2: Is this really a sub-step? If it is, then consider making the previous command in this step also a sub-step.

Author: This is a substep to the Optional adding AbortOnFailure action

+
[source,terminal]
----
$ oc get ibgu -o yaml
----

> . Add the action to your existing `ImageBasedGroupUpgrade` CR by running the following command:

Reviewer 2: `Add the action`--which action?
